### PR TITLE
Implement new features

### PR DESCRIPTION
**Pull request by orvend**
Add new functionalities for basic descriptive statistical analysis

### CODE REVIEW

File: stats_test.go [hunk 1]
```diff
@@ -109,3 +109,123 @@ func BenchmarkMedian2(b *testing.B)   { benchmarkMean(2, b) }
...
+		input []float64
+	}
+	tests := []struct {
+		name string
+		args args
+		want float64
+	}{
+		{"empty case", args{[]float64{}}, math.NaN()},
+		{"equal case", args{[]float64{5.0, 5.0, 5.0, 5.0}}, 5.0},
+		{"normal case", args{[]float64{1.0, 5.0, 2.0, 6.0, 1.0, 2.0, 3.0}}, 6.0},
```
Comment (qmuntal) [inline stats_test.go]: missing case with a NaN in the input

File: stats_test.go [hunk 2]
```diff
@@ -109,3 +109,123 @@ func BenchmarkMedian2(b *testing.B)   { benchmarkMean(2, b) }
...
+		input []float64
+	}
+	tests := []struct {
+		name string
+		args args
+		want float64
+	}{
+		{"empty case", args{[]float64{}}, math.NaN()},
+		{"equal case", args{[]float64{5.0, 5.0, 5.0, 5.0}}, 5.0},
+		{"normal case", args{[]float64{1.0, 5.0, 2.0, 6.0, 1.0, 2.0, 3.0}}, 1.0},
```
Comment (qmuntal) [inline stats_test.go]: missing case with a NaN in the input

File: stats_test.go [hunk 3]
```diff
@@ -109,3 +109,123 @@ func BenchmarkMedian2(b *testing.B)   { benchmarkMean(2, b) }
...
+		input []float64
+	}
+	tests := []struct {
+		name string
+		args args
+		want float64
+	}{
+		{"empty case", args{[]float64{}}, math.NaN()},
+		{"equal case", args{[]float64{5.0, 5.0, 5.0, 5.0}}, 0.0},
+		{"normal case", args{[]float64{1.0, 5.0, 2.0, 6.0, 1.0, 2.0, 3.0}}, 5.0},
```
Comment (qmuntal) [inline stats_test.go]: missing case with a NaN in the input

File: stats_test.go [hunk 4]
```diff
@@ -109,3 +109,123 @@ func BenchmarkMedian2(b *testing.B)   { benchmarkMean(2, b) }
...
+		input []float64
+	}
+	tests := []struct {
+		name string
+		args args
+		want float64
+	}{
+		{"empty case", args{[]float64{}}, math.NaN()},
+		{"single value case", args{[]float64{1.0}}, math.NaN()},
+		{"normal case", args{[]float64{1.0, 2.0, 3.0, 2.3, 1.4, 1.7, 1.5, 1.5, 1.8, 2.6, 2.3, 2.0, 2.2}}, 0.542548},
```
Comment (qmuntal) [inline stats_test.go]: missing case with a NaN in the input

File: stats_test.go [hunk 5]
```diff
@@ -109,3 +109,123 @@ func BenchmarkMedian2(b *testing.B)   { benchmarkMean(2, b) }
...
+		input []float64
+	}
+	tests := []struct {
+		name string
+		args args
+		want float64
+	}{
+		{"empty case", args{[]float64{}}, math.NaN()},
+		{"single value case", args{[]float64{1.0}}, math.NaN()},
+		{"normal case", args{[]float64{1.0, 2.0, 3.0, 2.3, 1.4, 1.7, 1.5, 1.5, 1.8, 2.6, 2.3, 2.0, 2.2}}, 0.294358},
```
Comment (qmuntal) [inline stats_test.go]: missing case with a NaN in the input

File: stats.go [hunk 6]
```diff
@@ -34,3 +34,109 @@ func Median(input []float64) (output float64) {
...
+	mean := Mean(input)
+	ssd := 0.0
+	for _, o := range input {
+		ssd += math.Pow(o-mean, 2.0)
+	}
+	return ssd
+}
+
+// StandardDeviationSample returns the standard deviation of the sample
+func StandardDeviationSample(input []float64) float64 {
```
Comment (qmuntal) [inline stats.go]: Why adding `Sample` at the end? I think `StdDev` is enough and generally understood.

File: stats.go [hunk 7]
```diff
@@ -34,3 +34,109 @@ func Median(input []float64) (output float64) {
...
+	return math.Sqrt(VarianceSample(input))
+}
+
+// VarianceSample returns the variance of the sample
+func VarianceSample(input []float64) float64 {
```
Comment (qmuntal) [inline stats.go]: Why adding `Sample` at the end? I think `Variance` is enough

File: stats.go [hunk 8]
```diff
@@ -34,3 +34,109 @@ func Median(input []float64) (output float64) {
...
+	num := sumOfSquaredDifferences(input)
+	den := float64(len(input) - 1)
+	return num / den
```
Comment (qmuntal) [inline stats.go]: Is it faster if you call the two parts inline, like `return sumOfSquaredDifferences(input) / float64(len(input) - 1)`?

File: stats.go (hunk 6 above)
Comment (orvend) [reply stats.go]: Because the formula is different if you calculate the StdDev of a Sample or of a Population

Comment (orvend) [reply stats.go]: I wanted also to implement StandardDeviationPopulation

Comment (qmuntal) [reply stats.go]: I would assume that if nothing else is specified the `StdDev` is the sampled versio, and I think that 99% of people using this function are also expecting it.
If later you want to add the Population version you can always add this keyword to the end of the name.

File: stats.go (hunk 8 above)
Comment (orvend) [reply stats.go]: Yes! Performance is improved.
From:
```
BenchmarkVarianceSample0-32             1000000000               3.20 ns/op            0 B/op          0 allocs/op
BenchmarkVarianceSample1-32             1000000000               2.91 ns/op            0 B/op          0 allocs/op
BenchmarkVarianceSample2-32             69475694                86.8 ns/op             0 B/op          0 allocs/op
BenchmarkVarianceSample10-32            13512231               405 ns/op               0 B/op          0 allocs/op
BenchmarkVarianceSample1e3-32             120768             49563 ns/op               0 B/op          0 allocs/op
BenchmarkVarianceSample1e6-32                118          52834614 ns/op               0 B/op          0 allocs/op
PASS
ok      github.com/orvend/stats 100.480s
```
To:
```
BenchmarkVarianceSample0-32             1000000000               2.81 ns/op            0 B/op          0 allocs/op
BenchmarkVarianceSample1-32             1000000000               2.75 ns/op            0 B/op          0 allocs/op
BenchmarkVarianceSample2-32             76150934                77.6 ns/op             0 B/op          0 allocs/op
BenchmarkVarianceSample10-32            15028654               376 ns/op               0 B/op          0 allocs/op
BenchmarkVarianceSample1e3-32             120529             46184 ns/op               0 B/op          0 allocs/op
BenchmarkVarianceSample1e6-32                128          49593414 ns/op               0 B/op          0 allocs/op
PASS
ok      github.com/orvend/stats 98.807s
```

File: stats.go (hunk 6 above)
Comment (orvend) [reply stats.go]: I agree with you. Sampled version is the most common use.

File: stats.go (hunk 7 above)
Comment (orvend) [reply stats.go]: Ok

File: stats_test.go (hunk 4 above)
Comment (orvend) [reply stats_test.go]: Ok

**Pull request merged**